## **IST-718** **Big Data Analytics**

---



### **Project Title**

**Title**- New Zealand Airline's Air-Fare Prediction

**Project By:**






> **Jasdeep Monga**-414509580

> **Priya Rao**-660997227

> **Yunhuai Cao**-358813275

**Group Number** - 19





### **Abstract**

**Overview**- To predict Domestic Airfare of New-Zealand Airlines basd on the given predictors.

**Data Set Description**- We have **162833 rows** and **11 columns**.

**Dataset link**: [NZ_Airfares](https://drive.google.com/file/d/1S3OBGMwh-FhYIlIpCeAt9afSXYrZghXa/view?usp=share_link)

**Interesting Facts**:

1.   The data was collected by **Expedia Group**, one of the biggest online travel company and it is from year **2019**.
2.   Major Airlines in New Zealand are **Air New Zealand**, **JetStar** and **SoundsAir**.
3.   Data ranges from **19th September 2019** to **18th December 2019**, a total of **90 days**.
4.   From the **Travel date** column we were able to extract the **Day**,**Month** and **Quarter**.
5.   The flight tickets for all the three airlines were cheaper for the month of december (holiday season).

**Specific Predictions**:

1.   By comparing the **airfare** across airlines we were able to compare the cheapest and the expensive airlines.
2.   With **Airfare** we also checked how the fare changes for flights with no stops i.e. direct flights vs the ones with multiple stops.
3. We were able to find the particular **Day** and the **Month** that is best suitable for  travel.


**Specific Inferences**:

1.   **Air New Zealand** is the expensive airline with **\$418.79** and **Jetstar** is the **cheapeast** i.e. **\$165.1**.
2.   **Jetstar** is the **cheapest** on **Direct** and **1 stop routes** whereas **Air New Zealand** is **expensive** all across .
3. In **Novemeber** airline tickets were expensive 
i.e. **\$421.74** and tickets were cheaper in **December**
i.e. **\$365.03**.

**Concusion Summary**

As per the results we can conclude that **JetStar airline** would be the 1st choice to commute to any place within New Zealand because the average fare is around **\$165.1**.We must schedule our travel for the month of **December**, and we must do so on **weekdays**, including **Tuesday** and **Wednesday**, when the cost will be lower than on weekends.







In [1]:
# Installing pyspark
!pip install pyspark py4j

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 281.4 MB 26 kB/s 
     |████████████████████████████████| 200 kB 44.3 MB/s 
     |████████████████████████████████| 199 kB 10.8 MB/s 
  Created wheel for pyspark: filename=pyspark-3.3.1-py2.py3-none-any.whl size=281845512 sha256=28e823e7fb453c0ecea60af824eca8024c60f2f89004f2bc0a472727e36bf9c2
  Stored in directory: /root/.cache/pip/wheels/43/dc/11/ec201cd671da62fa9c5cc77078235e40722170ceba231d7598
Successfully built pyspark


In [2]:
%%bash
# Do not change or modify this cell
# Need to install pyspark
# if pyspark is already installed, will print a message indicating pyspark already installed
# pip install pyspark &> /dev/null

# Download the data files from github
# If the data file does not exist in the colab environment
data_file_1=NZ_Airfares.csv

if [[ ! -f ./${data_file_1} ]]; then 
   # download the data file from github and save it in this colab environment instance
   wget https://raw.githubusercontent.com/jasumonga17/New-Zealand-AirFare/main/NZ_Airfares.csv
fi

--2022-12-14 05:01:09--  https://raw.githubusercontent.com/jasumonga17/New-Zealand-AirFare/main/NZ_Airfares.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 13271399 (13M) [text/plain]
Saving to: ‘NZ_Airfares.csv’

     0K .......... .......... .......... .......... ..........  0% 4.08M 3s
    50K .......... .......... .......... .......... ..........  0% 5.89M 3s
   100K .......... .......... .......... .......... ..........  1% 22.7M 2s
   150K .......... .......... .......... .......... ..........  1% 17.9M 2s
   200K .......... .......... .......... .......... ..........  1% 8.00M 2s
   250K .......... .......... .......... .......... ..........  2% 32.0M 1s
   300K .......... .......... .......... .......... ..........  2% 79.1M 1s
   350K ..........

In [3]:
# Importing the Necesaary libraries
from pyspark.sql import SparkSession
from pyspark.sql import Row
from pyspark.sql import functions as fn
import matplotlib.pyplot as plt

In [4]:
# Initiating Spark Session
# appName is the name of the application, which will be displayed in the spark UI.
# getOrCreate() will either create an existing SparkSession or, if none exists, will create one.
spark=SparkSession.builder.appName('spark-intro').getOrCreate()
sc=spark.sparkContext

In [5]:
# Reading the CSV file
# inferSchema- is to understand automatically that the data is string or integer or double.
data=spark.read.csv('NZ_Airfares.csv',header=True,inferSchema=True)

In [6]:
data.show(5)

+-----------+------------+---------+------------+---------+--------+--------+-------+-------+---------------+------------+
|Travel Date|Dep. airport|Dep. time|Arr. airport|Arr. time|Duration|  Direct|Transit|Baggage|        Airline|Airfare(NZ$)|
+-----------+------------+---------+------------+---------+--------+--------+-------+-------+---------------+------------+
|    9/19/19|         AKL|  1:35 PM|         CHC|  3:00 PM|  1h 25m|(Direct)|   null|   null|        Jetstar|         111|
|    9/19/19|         AKL|  3:55 PM|         CHC|  5:20 PM|  1h 25m|(Direct)|   null|   null|        Jetstar|         111|
|    9/19/19|         AKL| 11:40 AM|         CHC|  1:05 PM|  1h 25m|(Direct)|   null|   null|        Jetstar|         132|
|    9/19/19|         AKL|  8:00 PM|         CHC|  9:25 PM|  1h 25m|(Direct)|   null|   null|        Jetstar|         132|
|    9/19/19|         AKL|  9:00 AM|         CHC| 10:25 AM|  1h 25m|(Direct)|   null|   null|Air New Zealand|         133|
+-----------+---

In [7]:
# Schema
# printSchema()-gives what kind of fields are in the dataset.
print("Schema:", data.printSchema())

root
 |-- Travel Date: string (nullable = true)
 |-- Dep. airport: string (nullable = true)
 |-- Dep. time: string (nullable = true)
 |-- Arr. airport: string (nullable = true)
 |-- Arr. time: string (nullable = true)
 |-- Duration: string (nullable = true)
 |-- Direct: string (nullable = true)
 |-- Transit: string (nullable = true)
 |-- Baggage: string (nullable = true)
 |-- Airline: string (nullable = true)
 |-- Airfare(NZ$): integer (nullable = true)

Schema: None


In [8]:
# Shape of DataFrame
print("Rows:",data.count(),"Columns:",len(data.columns))

Rows: 162833 Columns: 11


In [9]:
# Updated DataFrame
data_df=data.toPandas()

In [10]:
# Pandas DataFrame
data_df.head()

Travel Date Dep. airport Dep. time Arr. airport Arr. time Duration  \
0     9/19/19          AKL   1:35 PM          CHC   3:00 PM   1h 25m   
1     9/19/19          AKL   3:55 PM          CHC   5:20 PM   1h 25m   
2     9/19/19          AKL  11:40 AM          CHC   1:05 PM   1h 25m   
3     9/19/19          AKL   8:00 PM          CHC   9:25 PM   1h 25m   
4     9/19/19          AKL   9:00 AM          CHC  10:25 AM   1h 25m   

     Direct Transit Baggage          Airline  Airfare(NZ$)  
0  (Direct)    None    None          Jetstar           111  
1  (Direct)    None    None          Jetstar           111  
2  (Direct)    None    None          Jetstar           132  
3  (Direct)    None    None          Jetstar           132  
4  (Direct)    None    None  Air New Zealand           133

### **Data Cleaning**




1. In our data set one of the major challenge was to set the right format for '**Travel Date**' column.

2. Data that we got from the source had **80,000 records** in **string format** and the rest **80,000 records** in **date format**.

3. In order to fix the date column we first had to set the right format for all the string records using the in-built pyspark functions and then set all records in one date format i.e. ***YYYY-MM-DD***.

4. The above step helped us in extracting **Day**, **Month** and **Quarter** from our **Travel date** column.

5. **Handling Missing Values**:

  *   **Departure Airport**-24
  *   **Arrival Airport**-24
  *   **Arrival Time**-5
  *   **Transit**-39756
  *   **Baggage**-160522
  *   **Airline**-5

6. For the 1st **3** attributes i.e.(**Departure Airport**,**Arrival Airport** and **Arrival Time**) we dropped those **53 records**.

7. For the other **3** attributes(**Transit**,**Baggage** and **Airline**) we **imputed** the missing values with **No Transit** for **Transit** column, **Not defined** for **Baggage** column and **Other** for **Airline** wherever the information was missing.

8. We set the name of the columns in readable format so that the **notations** are easy to understand i.e.:

  *   Dep. Airport - **Departure Airport**
  *   Arr. Airport - **Arrival Airport**
  *   Arr. Time -  **Arrival Time**
  *   Airfare(NZ$) - **Airfare(NZ)**

9. We changed the notation for **airport names** as well for better understanding :

  *   CHC - **ChristChurch**
  *   WLG - **Wellington**
  *   AKL - **Auckland**
  *   ZQN - **Queenstown**
  *   NPL - **New Plymouth**
  *   DUD - **Dunedin**
  *   NPE - **Hawke's**
  *   NSN - **Nelson**
  *   PMR- **Palmerston**

10. We created a new column **Average_Fare** i.e., wherever the Average Fare price was more than normal fare we gave it a value of **1** and **else** it would be **0**.

11. Final shape of the dataframe after the data cleaning process was **162804 rows** and **14 columns**.

In [11]:
# Changing data type of Travel
import datetime 
import pandas as pd
dt=datetime.datetime.today()
data_df['Travel Date']=pd.to_datetime(data_df['Travel Date'])
data_df['Day']=data_df['Travel Date'].dt.day_name()
data_df['Month']=data_df['Travel Date'].dt.strftime('%B')
# New column Quarter 
data_df['Quarter']=data_df['Travel Date'].dt.quarter
# Direct
data_df['Direct'] = data_df['Direct'].str.strip('()')

In [12]:
# Spark data frame
data_spark_df=spark.createDataFrame(data_df)
data_spark_df.show(5)

+-------------------+------------+---------+------------+---------+--------+------+-------+-------+---------------+------------+--------+---------+-------+
|        Travel Date|Dep. airport|Dep. time|Arr. airport|Arr. time|Duration|Direct|Transit|Baggage|        Airline|Airfare(NZ$)|     Day|    Month|Quarter|
+-------------------+------------+---------+------------+---------+--------+------+-------+-------+---------------+------------+--------+---------+-------+
|2019-09-19 00:00:00|         AKL|  1:35 PM|         CHC|  3:00 PM|  1h 25m|Direct|   null|   null|        Jetstar|         111|Thursday|September|      3|
|2019-09-19 00:00:00|         AKL|  3:55 PM|         CHC|  5:20 PM|  1h 25m|Direct|   null|   null|        Jetstar|         111|Thursday|September|      3|
|2019-09-19 00:00:00|         AKL| 11:40 AM|         CHC|  1:05 PM|  1h 25m|Direct|   null|   null|        Jetstar|         132|Thursday|September|      3|
|2019-09-19 00:00:00|         AKL|  8:00 PM|         CHC|  9:25 

In [13]:
# Data Pre-Processing
from pyspark.sql import functions as fn
from pyspark.sql.functions import col
from pyspark.sql.types import IntegerType,BooleanType,DateType,LongType
from pyspark.sql.functions import *
from pyspark.sql.types import StructField
from pyspark.sql.functions import expr

In [14]:
# Changing Column Names
data_updated=data_spark_df.withColumnRenamed("Dep. airport","Departure_Airport").withColumnRenamed("Dep. time","Departure_Time").withColumnRenamed("Arr. airport","Arrival_Airport").withColumnRenamed("Arr. time","Arrival_Time").withColumnRenamed("Airfare(NZ$)","Airfare(NZ)").withColumnRenamed('Travel Date','Travel_Date')
data_updated.show(5)

+-------------------+-----------------+--------------+---------------+------------+--------+------+-------+-------+---------------+-----------+--------+---------+-------+
|        Travel_Date|Departure_Airport|Departure_Time|Arrival_Airport|Arrival_Time|Duration|Direct|Transit|Baggage|        Airline|Airfare(NZ)|     Day|    Month|Quarter|
+-------------------+-----------------+--------------+---------------+------------+--------+------+-------+-------+---------------+-----------+--------+---------+-------+
|2019-09-19 00:00:00|              AKL|       1:35 PM|            CHC|     3:00 PM|  1h 25m|Direct|   null|   null|        Jetstar|        111|Thursday|September|      3|
|2019-09-19 00:00:00|              AKL|       3:55 PM|            CHC|     5:20 PM|  1h 25m|Direct|   null|   null|        Jetstar|        111|Thursday|September|      3|
|2019-09-19 00:00:00|              AKL|      11:40 AM|            CHC|     1:05 PM|  1h 25m|Direct|   null|   null|        Jetstar|        132|Th

In [15]:
# Null Value check in the dataset
print("Schema:", data_updated.printSchema())

root
 |-- Travel_Date: timestamp (nullable = true)
 |-- Departure_Airport: string (nullable = true)
 |-- Departure_Time: string (nullable = true)
 |-- Arrival_Airport: string (nullable = true)
 |-- Arrival_Time: string (nullable = true)
 |-- Duration: string (nullable = true)
 |-- Direct: string (nullable = true)
 |-- Transit: string (nullable = true)
 |-- Baggage: string (nullable = true)
 |-- Airline: string (nullable = true)
 |-- Airfare(NZ): long (nullable = true)
 |-- Day: string (nullable = true)
 |-- Month: string (nullable = true)
 |-- Quarter: long (nullable = true)

Schema: None


In [16]:
#Checking for Null Values in all the columns
from pyspark.sql.functions import col,isnan, when, count
data_updated.select([count(when(isnan(c) | col(c).isNull(), c)).alias(c) for c in data_updated.columns[1:]]
   ).show()


+-----------------+--------------+---------------+------------+--------+------+-------+-------+-------+-----------+---+-----+-------+
|Departure_Airport|Departure_Time|Arrival_Airport|Arrival_Time|Duration|Direct|Transit|Baggage|Airline|Airfare(NZ)|Day|Month|Quarter|
+-----------------+--------------+---------------+------------+--------+------+-------+-------+-------+-----------+---+-----+-------+
|               24|             0|             24|           5|       0|     0|  39756| 160522|      5|          0|  0|    0|      0|
+-----------------+--------------+---------------+------------+--------+------+-------+-------+-------+-----------+---+-----+-------+



In [17]:
# Dropping Records from Departure Airport and Arrival Airport
data_updated=data_updated.na.drop(subset=['Departure_Airport','Arrival_Airport','Arrival_Time'])

In [18]:
# Replacing Null Values in Transit, Baggage and Airline Column
data_updated=data_updated.na.fill('No Transit',['Transit']).na.fill('Not Defined',['Baggage']).na.fill('Other',['Airline'])

In [19]:
# Modifying the airport names
data_updated=data_updated.replace('CHC','ChristChurch').replace('WLG','Wellington').replace('AKL','Auckland').replace('ZQN','Queenstown').replace('NPL','New Plymouth Airport').replace('DUD','Dunedin').replace('NPE',"Hawke's").replace('NSN','Nelson').replace('PMR','Palmerston')

In [20]:
# Shape of DataFrame
print("Rows:",data_updated.count(),"Columns:",len(data_updated.columns))

Rows: 162804 Columns: 14


In [21]:
#Checking for Null Values in all the columns
from pyspark.sql.functions import col,isnan, when, count
data_updated.select([count(when(isnan(c) | col(c).isNull(), c)).alias(c) for c in data_updated.columns[1:]]
   ).show()

+-----------------+--------------+---------------+------------+--------+------+-------+-------+-------+-----------+---+-----+-------+
|Departure_Airport|Departure_Time|Arrival_Airport|Arrival_Time|Duration|Direct|Transit|Baggage|Airline|Airfare(NZ)|Day|Month|Quarter|
+-----------------+--------------+---------------+------------+--------+------+-------+-------+-------+-----------+---+-----+-------+
|                0|             0|              0|           0|       0|     0|      0|      0|      0|          0|  0|    0|      0|
+-----------------+--------------+---------------+------------+--------+------+-------+-------+-------+-----------+---+-----+-------+



In [22]:
# Creating Average Fare column 
# If Avg Fare>Fare then 1 else 0
data_updated=data_updated.withColumn('Average_Fare',when(fn.lit(data_updated.approxQuantile('Airfare(NZ)',[0.5],0)[0])>col('Airfare(NZ)'),1).otherwise(0))


In [23]:
df=data_updated.toPandas()

### **Data Exploration Insights**


* For data exploration we have used the **Plotly** library.

**Airline vs Average Fare** 

---


1.  We plotted a bar graph to see the Average Airline fare across each Airline in New Zealand.
2.  **Air New Zealand** is the costliest airline with **\$418.79** and **Jetstar** is the **cheapeast** i.e. **\$165.1**.









In [24]:
# Converting into pandas DataFrame
import pandas as pd
from pyspark.sql.functions import sum,avg,max,mean

# Converting into Pandas Data Frame
mean_fare=data_updated.groupby('Airline').agg(fn.round(mean('Airfare(NZ)'),2).alias('Average_Fare'))

# Creating pandas dataframe
mean_fare_df=mean_fare.toPandas()
mean_fare_df=mean_fare_df.sort_values(by='Average_Fare',ascending=False)

In [25]:
import plotly.express as px
fig=px.bar(mean_fare_df,x='Airline',y='Average_Fare',color='Airline',text_auto=True)
fig.update_layout(title_text="Airline's Average Fare")
fig.show()

**Direct vs Average Fare** 

---
1.  We plotted a **stacked** **bar** graph to see the **Average Airline Fare** across the **flight route**.
2. To see how fare changes if they airline has no stops or stops in between.
3.  **Air New Zealand** is an  expesnive airline with  an average fare of **\$236.78** on **Direct route** and **Jetstar** is the **cheapeast** i.e. **\$128.16** on **Direct** and **1 stop routes**.
4. **SoundsAir** and **JetStar** don't operate on routes having **3 stops**.


In [26]:
# Converting into Pandas Data Frame
direct_fare=data_updated.groupby('Airline','Direct').agg(fn.round(mean('Airfare(NZ)'),2).alias('Average_Fare'))

# Regex
from pyspark.sql.functions import regexp_replace

# Creating pandas dataframe
direct_fare_df=direct_fare.toPandas()
direct_fare_df=direct_fare_df.sort_values(by='Average_Fare',ascending=False)
direct_fare_df.head()


Airline   Direct  Average_Fare
0  Air New Zealand  2 stops        585.98
6  Air New Zealand  3 stops        553.30
4       Sounds Air  2 stops        553.27
2  Air New Zealand   1 stop        415.51
1       Sounds Air   1 stop        356.34

In [27]:
# Direct vs Average Fare
fig = px.bar(direct_fare_df, x="Direct", y="Average_Fare", color="Airline",text_auto=True)
fig.update_layout(title_text="Direct vs Average Fare")
fig.show()

**Month vs Average Fare** 

---
1.  We plotted a **line chart** to see the change in **Average Airline Fare** across each **Month**.

2.  In **November** the airfare was **most expensive** i.e **\$421.74** and in **December** the fare was **cheapest** i.e., **\$365.03**.

In [28]:
# Converting into Pandas Data Frame
month_fare=data_updated.groupby('Month').agg(fn.round(mean('Airfare(NZ)'),2).alias('Average_Fare'))

# Regex
from pyspark.sql.functions import regexp_replace

# Creating pandas dataframe
month_fare_df=month_fare.toPandas()
month_fare_df=month_fare_df.sort_values(by='Average_Fare',ascending=False)
month_fare_df.head()

Month  Average_Fare
2   November        421.74
0    October        417.15
1  September        413.40
3   December        365.03

In [29]:
# Line Chart
# Average Fare vs Month
fig = px.line(month_fare_df, x="Month", y="Average_Fare",text='Average_Fare')
fig.update_traces(textposition="bottom right")
fig.update_layout(title_text="Month vs Average Fare")
fig.show()

**Day vs Average Fare** 

---
1.  We plotted a **line chart** to see the change in **Average Airline Fare** across each **Day** in a Month.

2.  On **Weekends(Saturday and Sunday)** the airfare was **most expensive** i.e **\$484.50 and \$483.44** and on **Weekdays(Wedensday and Tuesday)** the fare was **cheapest** i.e., **\$365.71** and **\$365.66**. This is common around all **4 months**.

In [30]:
# Day and Month

# Converting into Pandas Data Frame
dm_fare=data_updated.groupby('Day','Month').agg(fn.round(mean('Airfare(NZ)'),2).alias('Average_Fare'))

# Regex
from pyspark.sql.functions import regexp_replace

# Creating pandas dataframe
dm_fare_df=dm_fare.toPandas()
dm_fare_df=dm_fare_df.sort_values(by='Average_Fare',ascending=False)
dm_fare_df.head()

Day      Month  Average_Fare
5   Saturday  September        484.50
8   Saturday    October        483.44
7     Sunday    October        480.74
27  Saturday   November        479.19
16    Sunday   November        469.11

In [31]:
# Line Chart
# Day vs Average Fare
fig = px.line(dm_fare_df, x="Day", y="Average_Fare",color='Month',text='Average_Fare',markers=True)
fig.update_traces(textposition="bottom right")
fig.update_layout(title_text="Day vs Average Fare")
fig.show()

### **Methodology**

1. In this section, we have performed the **feature engineering** i.e., used **String Indexing** on all the **categorical variables** and then stored them in a pipeline called **airfare_pipe**.

2. Next we performed the **Train, Validation** and **Test** **set** split i.e. **60:20:20**.

3. We created a **Vector Assembler** that stores all the values of predictor variables in a **single vector column** i.e., **features**.

4. **Standard Scaler** was applied on the features column that would bring all the features to the same scale.


In [32]:
# Pipeline
from pyspark.sql.functions import col, when
from pyspark.ml.pipeline import Pipeline, PipelineModel
from pyspark.ml.regression import LinearRegression
from pyspark.ml import feature

# Creating Airfare_pipe
airfare_pipe=Pipeline(stages=[feature.StringIndexer(inputCols=('Departure_Airport','Arrival_Airport','Direct','Transit','Baggage','Airline','Day','Month'),
                                                    outputCols=('encoded_Departure_Airport','encoded_Arrival_Airport','encoded_Direct','encoded_Transit','encoded_Baggage','encoded_Airline','enoded_Day','encoded_Month'))])

In [33]:
# Fitting Pipeline
air_fit=airfare_pipe.fit(data_updated)

# Transform
data_updated=air_fit.transform(data_updated)

# Dropping Columns
data_updated = data_updated.drop('Travel_Date','Departure_Time','Arrival_Time','Duration','Departure_Airport','Arrival_Airport','Direct','Transit','Baggage','Airline','Day','Month')

# Arranging the columns
data_updated = data_updated.select('Quarter','encoded_Departure_Airport','encoded_Arrival_Airport', 'encoded_Direct','encoded_Transit','encoded_Airline','enoded_Day','encoded_Month','Average_Fare','Airfare(NZ)')

In [34]:
# Train-Test Split
train, validation, test = data_updated.randomSplit([0.6, 0.2, 0.2],seed=0)
train.show(5)

+-------+-------------------------+-----------------------+--------------+---------------+---------------+----------+-------------+------------+-----------+
|Quarter|encoded_Departure_Airport|encoded_Arrival_Airport|encoded_Direct|encoded_Transit|encoded_Airline|enoded_Day|encoded_Month|Average_Fare|Airfare(NZ)|
+-------+-------------------------+-----------------------+--------------+---------------+---------------+----------+-------------+------------+-----------+
|      3|                      0.0|                    0.0|           0.0|            2.0|            0.0|       0.0|          2.0|           1|        307|
|      3|                      0.0|                    0.0|           0.0|            2.0|            0.0|       1.0|          2.0|           1|        307|
|      3|                      0.0|                    0.0|           0.0|            2.0|            0.0|       1.0|          2.0|           1|        307|
|      3|                      0.0|                    0.0

In [35]:
# Vetcor Assembler
va=feature.VectorAssembler(inputCols=['Quarter','encoded_Departure_Airport','encoded_Arrival_Airport', 'encoded_Direct','encoded_Transit','encoded_Airline','enoded_Day','encoded_Month','Average_Fare'], outputCol='features')

In [36]:
# Standard Scaler
sc=feature.StandardScaler(withMean=True,inputCol='features',outputCol='sc_features')

### **Model Prediction**

1. As per our data set our **target column** i.e. **Airfare(NZ)** is a continuous variable.

2. In order to predict the **Airfare** we used the following regression models:

  *   **Linear Reression**
  *   **Random Forest Regressor**
  *   **Gradient Boosting Regressor**

####  **1. Linear Regression**

*   For **Linear Regression Model**, we have created a separate pipeline called **lr_pipe**.
*   Inside the **lr_pipe** we made use of **Vector Assembler**, **Standard Scaler** and **Linear regression object.**
*   We performed a **fit** on the **train set** and **transformed** on the **validation** and **test set**.
*   We obtained the **MSE** and the **R2-squared value**.





In [37]:
# Model Building

# Linear Regression
from pyspark.sql.functions import col, when
from pyspark.ml.pipeline import Pipeline, PipelineModel
from pyspark.ml.regression import LinearRegression
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml import feature

# Regression object
lr=LinearRegression(featuresCol='sc_features',labelCol='Airfare(NZ)')

# Pipeline
lr_pipeline=Pipeline(stages=[va,sc,lr])

# Fit on Train set
train_lr=lr_pipeline.fit(train)

# Validation Set
val_lr=train_lr.transform(validation)

# Test set
test_lr=train_lr.transform(test)

evaluator = RegressionEvaluator(labelCol='Airfare(NZ)',predictionCol='prediction')
lr_validation_mse=evaluator.evaluate(val_lr,{evaluator.metricName: "mse"})
print('MSE_LR_Validation:',lr_validation_mse)
lr_test_mse=evaluator.evaluate(test_lr,{evaluator.metricName: "mse"})
print('MSE_LR_Test:',lr_test_mse)
print('R2_LR:',evaluator.evaluate(val_lr,{evaluator.metricName: "r2"}))

MSE_LR_Validation: 11112.697940423024
MSE_LR_Test: 10971.035526181646
R2_LR: 0.6110991169929199


#### **2. Random Forest Regressor**

*   For **Random Forest Regressor Model**, we have created a separate pipeline called **rf_pipe**.
*   Inside the **rf_pipe** we made use of **Vector Assembler**, **Standard Scaler** and **Random Forest Regression object.**
*   We performed a **fit** on the **train set** and **transformed** on the **validation** and **test set**.
*   We obtained the **MSE** and the **R2-squared value**.

In [38]:
# Random Forest Regressor
from pyspark.ml.regression import RandomForestRegressor
from pyspark.ml.evaluation import RegressionEvaluator

# Regression object
rf=RandomForestRegressor(featuresCol='sc_features',labelCol='Airfare(NZ)',maxBins=1221)

# Pipeline
rf_pipeline=Pipeline(stages=[va,sc,rf])

# Fit on Train set
train_rf=rf_pipeline.fit(train)

# Validation Set
val_rf=train_rf.transform(validation)

# Test set
test_rf=train_rf.transform(test)

evaluator = RegressionEvaluator(labelCol='Airfare(NZ)',predictionCol='prediction')
print('MSE_RF_Validation:',evaluator.evaluate(val_rf,{evaluator.metricName: "mse"}))
rf_test_mse=evaluator.evaluate(test_rf,{evaluator.metricName: "mse"})
print('MSE_LR_Test:',rf_test_mse)
print('R2_RF:',evaluator.evaluate(val_rf,{evaluator.metricName: "r2"}))

MSE_RF_Validation: 9127.375095754001
MSE_LR_Test: 9048.608370579706
R2_RF: 0.6805776371043484


##### **A.) Cross Validation-Random Forest Regressor**

*   In order to find the **best model parameters** we used **3 Fold Cross Validation**

*   In case of **Random Forest** we defined the 2 features i.e., **numTrees-** **Number of Trees** and **maxDepth** i.e., the **maximum depth** at which we want the tree to grow post that there would be no splitting.

*  We gave two values i.e., **5** and **10** for both the features. 
*   Our best model gave us **10** as the **best value** for both features.





In [39]:
# Cross-Validation Random Forest Regressor

from pyspark.ml.tuning import ParamGridBuilder
import numpy as np
from pyspark.ml.tuning import CrossValidator
paramGrid = ParamGridBuilder() \
.addGrid(rf.numTrees, [5,10]) \
.addGrid(rf.maxDepth, [5,10]) \
.build()
crossval = CrossValidator(estimator=rf_pipeline,
                          estimatorParamMaps=paramGrid,
                          evaluator=RegressionEvaluator(labelCol='Airfare(NZ)',predictionCol='prediction'),
                          numFolds=3)
cvModel=crossval.fit(train)
predictions=cvModel.transform(validation)
best=cvModel.bestModel
bestModel = best.stages[2]
print('numTrees - ', bestModel.getNumTrees)
print('maxDepth - ', bestModel.getOrDefault('maxDepth'))


numTrees -  10
maxDepth -  10


##### **B.) Best Model-Random Forest**

*   After obtaining the **best model parameters** we substituted those values in the **RandomForestRegressor()**.
*   We mentioned the **numTrees=10** and **maxDepth=10** this would give us the best possible **MSE** and **R2-squred value** for Random Forest Model. The values obtained were as follows:

  *   **RF_Validation_MSE**: 5904.878103737984
  *   **RF_Test_MSE**: 5865.389501246651
  *   **R2_RF**: 0.793352404528197

*  We can also observe the important features that have the maximum importance in achieveing such results.

* **Average_Fare** has an **importance of 59.7%** followed by **Transit-14.64%** etc.



In [40]:
# Best Model-Random Forest
rf_best=RandomForestRegressor(numTrees=10, maxDepth=10,labelCol='Airfare(NZ)',predictionCol='prediction',maxBins=1221)
best_rf_pipe=Pipeline(stages=[va,sc,rf_best])

# Train
train_rfe=best_rf_pipe.fit(train)

# Validation
val_rfe=train_rfe.transform(validation)

# Test Data
test_rfe=train_rfe.transform(test)

#MSE Scoring Metric
evaluator = RegressionEvaluator(labelCol='Airfare(NZ)',predictionCol='prediction')
rf_validation_mse=evaluator.evaluate(val_rfe,{evaluator.metricName: "mse"}) 
#rf_train_mse=evaluator.evaluate(train_rfe_transform,{evaluator.metricName: "mse"}) 
rf_test_mse=evaluator.evaluate(test_rfe,{evaluator.metricName: "mse"}) 
print('RF_Validation_MSE:',rf_validation_mse)
print('RF_Test_MSE:',rf_test_mse)
print('R2_RF:',evaluator.evaluate(val_rfe,{evaluator.metricName: "r2"}))

RF_Validation_MSE: 6341.579941692227
RF_Test_MSE: 6285.657708364491
R2_RF: 0.7780695514081243


In [41]:
# Feature Importance-Random Forest
import pandas as pd
best_model=train_rfe.stages[-1]

#best_model.featureImportances.toArray()

rf_feature_importance=pd.DataFrame(list(zip(data_updated.columns, best_model.featureImportances.toArray())),
            columns = ['feature', 'importance']).sort_values('importance',ascending=False)

display(rf_feature_importance)

feature  importance
8               Average_Fare    0.620747
3             encoded_Direct    0.137668
4            encoded_Transit    0.110471
2    encoded_Arrival_Airport    0.045324
6                 enoded_Day    0.032663
5            encoded_Airline    0.032508
1  encoded_Departure_Airport    0.017072
7              encoded_Month    0.003048
0                    Quarter    0.000498

#### **3. Gradient Boosting Regressor**

*   For **Gradient Boosting Regressor Model**, we have created a separate pipeline called **gb_pipeline**.
*   Inside the **gb_pipeline** we made use of **Vector Assembler**, **Standard Scaler** and **Gradient Boosting  object.**
*   We performed a **fit** on the **train set** and **transformed** on the **validation** and **test set**.
*   We obtained the **MSE** and the **R2-squared value**.

In [42]:
# GBT Regressor
from pyspark.ml.regression import GBTRegressor
from pyspark.ml.evaluation import RegressionEvaluator

# Regression object
gb=GBTRegressor(featuresCol='sc_features',labelCol='Airfare(NZ)',maxBins=1221)

# Pipeline
gb_pipeline=Pipeline(stages=[va,sc,gb])

# Fit on Train set
train_gb=gb_pipeline.fit(train)

# Validation Set
val_gb=train_gb.transform(validation)

# Test set
test_gb=train_gb.transform(test)

evaluator = RegressionEvaluator(labelCol='Airfare(NZ)',predictionCol='prediction')
print('MSE_GB_Validation:',evaluator.evaluate(val_gb,{evaluator.metricName: "mse"}))
gb_test_mse=evaluator.evaluate(test_gb,{evaluator.metricName: "mse"})
print('MSE_GB_Test:',gb_test_mse)
print('R2_GB:',evaluator.evaluate(val_gb,{evaluator.metricName: "r2"}))

MSE_GB_Validation: 7619.505151853937
MSE_GB_Test: 7523.448256659913
R2_GB: 0.7333471765794984


##### **A.) Cross Validation-GBT Regressor**

*   In order to find the **best model parameters** we used **3 Fold Cross Validation**

*   In case of **Gradient Boosting Regressor** we defined 2 features i.e., **maxIter-** **Number of Iterations** and **maxDepth** i.e., the **maximum depth** at which we want the tree to grow post that there would be no splitting.

*  We gave two values i.e., **5** and **10** for both the features 
*   Our best model gave us **10** as the best value for both features

In [43]:
# Cross Validation-GBT Regressor
from pyspark.ml.tuning import ParamGridBuilder
import numpy as np
from pyspark.ml.tuning import CrossValidator
paramGrid_gb = ParamGridBuilder() \
.addGrid(gb.maxIter, [5,10]) \
.addGrid(gb.maxDepth, [5,10]) \
.build()
crossval_gbt = CrossValidator(estimator=gb_pipeline,
                          estimatorParamMaps=paramGrid_gb,
                          evaluator=RegressionEvaluator(labelCol='Airfare(NZ)',predictionCol='prediction'),
                          numFolds=3)
cvModel_gbt=crossval_gbt.fit(train)
predictions=cvModel_gbt.transform(validation)
best_gbt=cvModel_gbt.bestModel
bestModel_gbt = best_gbt.stages[-1]
print('MaxIter=',bestModel_gbt._java_obj.getMaxIter())
print('MaxDepth=',bestModel_gbt._java_obj.getMaxDepth())


MaxIter= 10
MaxDepth= 10


##### **B.) Best Model-GB Regressor**

*   After obtaining the **best model parameters** we substituted those values in the **GradientBoostRegressor()**.
*   We mentioned the **maxIter=10** and **maxDepth=10** this would give us the best possible **MSE** and **R2-squred value** for Random Forest Model. The values obtained were as follows:

  *   **RF_Validation_MSE**: 6012.567430714048
  *   **RF_Test_MSE**: 5888.388311126792
  *   **R2_GB**: 0.7895837000627326

* **Average_Fare** has an **importance of 54%** followed by **Transit-23.8%** etc.



In [44]:
# Best Model-GB Regressor
gb_best=GBTRegressor(maxIter=10,maxDepth=10,labelCol='Airfare(NZ)',predictionCol='prediction',maxBins=1221)
best_gb_pipe=Pipeline(stages=[va,sc,gb_best])

# Train
train_gb=best_gb_pipe.fit(train)

# Validation
val_gb=train_gb.transform(validation)

# Test Data
test_gb=train_gb.transform(test)

#MSE Scoring Metric
evaluator = RegressionEvaluator(labelCol='Airfare(NZ)',predictionCol='prediction')
gbt_validation_mse=evaluator.evaluate(val_gb,{evaluator.metricName: "mse"}) 

#gbt_train_mse=evaluator.evaluate(train_gb_transform,{evaluator.metricName: "mse"}) 
gbt_test_mse=evaluator.evaluate(test_gb,{evaluator.metricName: "mse"}) 
#print(gbt_train_mse)
print('GB_Validation_MSE:',gbt_validation_mse)
print('GB_Test_MSE:',gbt_test_mse)
print('R2_GB:',evaluator.evaluate(val_gb,{evaluator.metricName: "r2"}))

GB_Validation_MSE: 6012.567430714048
GB_Test_MSE: 5888.388311126792
R2_GB: 0.7895837000627326


In [45]:
# Feature Importance Gradient Boosting Regressor

import pandas as pd
best_model=train_gb.stages[-1]

#best_model.featureImportances.toArray()

gb_feature_importance=pd.DataFrame(list(zip(data_updated.columns, best_model.featureImportances.toArray())),
            columns = ['feature', 'importance']).sort_values('importance',ascending=False)

display(gb_feature_importance)

feature  importance
8               Average_Fare    0.540219
4            encoded_Transit    0.238730
2    encoded_Arrival_Airport    0.073871
6                 enoded_Day    0.052895
1  encoded_Departure_Airport    0.036761
3             encoded_Direct    0.027493
7              encoded_Month    0.018884
5            encoded_Airline    0.007755
0                    Quarter    0.003393

#### **Model Inference**

1. As per the results obatined from all the Machine Learning models, **Gradient Boosting Regressor** has **performed the best** out of the lot. One major reason that it outperforms other models is that we are able to get **sequential models** by **combining weak learners into strong ones**.

2. It gave us an **model accuracy of 78.9%** that implies in business terms **78.9% change in the AirFare is being explained by change in all the predictor variables.**

3. The accuracy of the model can be increased beyond 80% by involving much more parameters in hyperparameter tuning but due the task being **computationally expensive** we had to restrict ourselves to **12 fits** in total i.e.**(3\* 2\*2=12)** 

4. In case of **GB Model** our **Airfare** of all the New Zeland airlines varies by **\$5k-$6k** in prediction.

5. **Average Fare** is the most **important feature** with an importance of **54%**. **Transit** is the **2nd** most **important feature** with an importance of **23.8%**.



  

#### **Conclusion**

1. We have a created a **dataframe** below that specifies the following:

  *   **Model Name**
  *   **Validation MSE**
  *   **Test MSE**
  *   **R-Squared Validation**

2. **Gradient Boosting Regressor** has **performed the best** out of the lot.

  * **MSE of the model** is as follows: 
  * **GB_Validation_MSE**: **6012.56**
  * **GB_Test_MSE**: **5888.388**

3. As per the **GB Model** our **Airfare** of all the **New Zeland airlines** varies by **\$5k-$6k** in prediction.

4. In case of **Linear Regression** and **Random Forest** the MSE is much higher and **R2-squared value** is lower i.e., in the range of **60-69%**.

5. **Average Fare** is the most important feature with an importance of **54%**. **Transit** is the **2nd** most important feature with an **importance of 23.8%**. This concludes that we should anlayze the **Average Fare** of an Airline over a period and also check if the flight has any **stopovers in between** before reaching the destination as that could impact the Airline Fare.

6. On **Weekends(Saturday and Sunday)** the airfare was **most expensive** i.e **\$484.50 and \$483.44** and on **Weekdays(Wedensday and Tuesday)** the fare was **cheapest** i.e., **\$365.71** and **\$365.66**. This is common around all **4 months**.

7. In **November** the airfare was **very expensive** i.e **\$421.74** and in **December** the fare was the **cheapest** i.e., **\$365.03**.

8. **Air New Zealand** is an expensive airline with an average fare of **\$236.78** on **Direct routes** and **Jetstar** is the **cheapeast** i.e. **\$128.16** on **Direct** and **1 stop routes**.

9. **SoundsAir** and **JetStar** don't operate on routes having **3 stops**.







In [46]:
# Creating a dataframe
import pandas as pd
lr_rf_gbt_mse_compare=pd.DataFrame({'Model':['LR','RF','GBT'],
                                 'Validation MSE':[lr_validation_mse,rf_validation_mse,gbt_validation_mse],
                                 'Test MSE':[lr_test_mse,rf_test_mse,gbt_test_mse],
                                 'R-Squared-Validation':[evaluator.evaluate(val_lr,{evaluator.metricName: "r2"}),
                                              evaluator.evaluate(val_rf,{evaluator.metricName: "r2"}),
                                              evaluator.evaluate(val_gb,{evaluator.metricName: "r2"})]
                                    
                                })
display(lr_rf_gbt_mse_compare)

Model  Validation MSE      Test MSE  R-Squared-Validation
0    LR    11112.697940  10971.035526              0.611099
1    RF     6341.579942   6285.657708              0.680578
2   GBT     6012.567431   5888.388311              0.789584